doc/specs/stdlib_hash_procedures.md

```diff
@@ -0,0 +1,1748 @@
+
+##### Syntax
+
+`code = [[stdlib_64_bit_hash_codes:pengy_hash]]( key, seed )`
+
+#####  Class
+
+Pure/elemental function
+
+##### Arguments
+
+`key`: shall be a scalar  expression of type default character or a
+Rank 1 integer vector expression of kind `INt8`, `INT16`, `INT32`, or
+`INTT64`. It is an `intent(in)` argument.
+
+`seed`: shall be an integer expression of kind `INT64`. It is
+an `intent(in)` argument.
+
+##### Result
+
+The result is an integer of kind `INT64`.
+
+##### Note
+
+`PENGY_HASH` is an implementation of the 64 bit `pengyhash` of Alberto
+Fajardo. The hash has acceptable performance on small keys, and good
+performance on long keys. It passes all the SMHasher tests, and has
+no known bad seeds.
+It is a *pure* function for integer arrays, and an *elemental*
+function for character strings. 
+
+##### Exampl
+
+```fortran
+    program demo_pengy_hash
+      use stdlib_64_bit_hash_codes, only: new_pengy_hash_seed, pengy_hash
+      use iso_fortran_env, only: int32, int64 
+      implicit none
+      integer, allocatable :: key(:)
+      integer(int64) :: hash
+      integer(int32)  ::  seed
+      key = [ 0, 1, 2, 3 ]
+      seed = 0_int32
+      call new_pengy_hash_seed( seed )
+      hash = pengy_hash( key, seed )
+      print *, seed, hash
+    end program demo_pengy_hash
+```
+
+
+#### `SPOOKY_HASH` - maps a character string or integer vector to an integer
+
+##### Status
+
+Experimental
+
+##### Description
+
+Maps a character string or integer vector to a 64 bit integer whose
+value also depends on a two element vector,  `seed`.
+
+##### Syntax
+
+`code = [[stdlib_64_bit_hash_codes:spooky_hash]]( key, seed )`
+
+#####  Class
+
+Function
+
+##### Arguments
+
+`key`: shall be a scalar of type default character expression or a
+Rank 1 integer vector expression of kind `INt8`, `INT16`, `INT32`, or
+`INTT64`. It is an `intent(in)` argument.
+
+`seed`: shall be a two element integer vector expression of kind
+`INT64`. It is an `intent(in)` argument.
+
+##### Result
+
+The result is a two element integer vector of kind `INT64`.
+
+##### Note
+
+`SPOOKY_HASH` is an implementation of the 64 bit version 2 of
+SpookyHash of Bob Jenkins. The code was designed for Little-Endian
+compilers. The output is different on Big Endian compilers, but still
+probably as good quality. It is often used as a 64 bit hash using the
+first element of the returned value, but can be used as a 128 bit
+hash. This version of `SPOOKY_HASH` has good performance on small keys
+and excellent performance on long keys. It passes all the SMHasher tests
+and has no known bad seeds.
+
+##### Example
+
+```fortran
+    program demo_spooky_hash
+      use stdlib_64_bit_hash_codes, only: new_spooky_hash_seed, &
+          spooky_hash
+      use iso_fortran_env, only: int64 
+      implicit none
+      integer, allocatable :: key(:)
+      integer(int64) :: hash(2), seed(2)
+      key = [ 0, 1, 2, 3 ]
+      seed = [ 119_int64, 2_int64**41-1 ]
```

Comment:
```suggestion
      seed = [ int64 :: 119, 2**41-1 ]
```
Personally, I find this form cleaner.

Reply:
This will not work properly.
It should be `seed = [integer(int64) :: 119, 2**41-1]`. However, the value `2**41-1` will be first computed with `integer(int32)` (which is out of range) and then be extended to `int64`.

Example:
```fortran
program  test
 use, intrinsic :: iso_fortran_env, only: int32, int64
 implicit none
 integer(int64) :: seed(2), seed1(2)

 seed = [119_int64, 2_int64**41-1]

 print*,seed

 seed1 = [integer(int64) :: 119, 2**41-1]

 print*,seed1
end program
```

Result with gfortran11:
```
                  119        2199023255551
                  119           2147483647
```

Reply:
Ouch. Clearly I've misunderstood the way the array initializer syntax works. Is there a list of Fortran caveat's where this can be documented?